core/templates/pages/exploration-player-page/services/stats-reporting.service.ts

```diff
@@ -16,344 +16,416 @@
+        exploration_version: this.explorationVersion,
+        state_name: stateName,
+        session_id: this.sessionId,
+        time_spent_in_state_secs: this.stateStopwatch.getTimeInSecs()
+      });
+    }
 
-        // Broadcast information about the state transition to listeners.
-        MessengerService.sendMessage(MessengerService.STATE_TRANSITION, {
-          explorationVersion: explorationVersion,
+    recordLeaveForRefresherExp(
+        stateName: string, refresherExpId: string): void {
+      this.currentStateName = stateName;
+      this.nextExpId = refresherExpId;
+      // TODO(#8038): Move this into a backend-api.service.
+      this.http.post(this.getFullStatsUrl('LEAVE_FOR_REFRESHER_EXP'), {
+        exploration_version: this.explorationVersion,
+        refresher_exp_id: refresherExpId,
+        state_name: stateName,
+        session_id: this.sessionId,
+        time_spent_in_state_secs: this.stateStopwatch.getTimeInSecs()
+      });
+    }
+
+    // Note that this also resets the stateStopwatch.
+    // The type of oldParams is declared as Object since it can vary depending
+    // on the oldStateName.
+    recordStateTransition(
+        oldStateName: string, newStateName: string, answer: string,
+        oldParams: Object, isFirstHit: boolean): void {
+      if (!this.aggregatedStats.state_stats_mapping.hasOwnProperty(
+        newStateName)) {
+        this.createDefaultStateStatsMapping(newStateName);
+      }
+      this.aggregatedStats.state_stats_mapping[
+        newStateName].total_hit_count += 1;
+      if (isFirstHit) {
+        this.aggregatedStats.state_stats_mapping[
+          newStateName].first_hit_count += 1;
+      }
+
+      this.previousStateName = oldStateName;
+      this.nextStateName = newStateName;
+      // TODO(#8038): Move this into a backend-api.service.
+      this.http.post(this.getFullStatsUrl('STATE_HIT'), {
+        // This is the time spent since the last submission.
+        client_time_spent_in_secs: this.stateStopwatch.getTimeInSecs(),
+        exploration_version: this.explorationVersion,
+        new_state_name: newStateName,
+        old_params: oldParams,
+        session_id: this.sessionId,
+      });
+
+      // Broadcast information about the state transition to listeners.
+      this.messengerService.sendMessage(
+        this.messengerService.STATE_TRANSITION, {
+          explorationVersion: this.explorationVersion,
           jsonAnswer: JSON.stringify(answer),
           newStateName: newStateName,
           oldStateName: oldStateName,
           paramValues: oldParams
         });
 
-        if (!statesVisited.hasOwnProperty(newStateName)) {
-          statesVisited[newStateName] = true;
-          numStatesVisited++;
-          SiteAnalyticsService.registerNewCard(numStatesVisited);
-        }
+      if (!this.statesVisited.has(newStateName)) {
+        this.statesVisited.add(newStateName);
+        this.numStatesVisited++;
```

Comment:
We shouldn't need this property anymore, we can just use `this.statesVisited.size` instead.

Reply:
Done.